### DISCUSSION TITLE
単体テストの考え方/使い方 10章

### DISCUSSION THREAD

**Discussion by y-kaneda** [atman-inc/training#105] [General]
輪読会のサマリ用ディスカッション。
10章分の各章単位程度でコメントを記載するので、不明点、相談点、確認点、勉強になったポイントなどは、各自ディスカッションのスレッドへコメントをお願いします。

**Comment by y-kaneda** [comment]
# 第10章 データベースに対するテスト

**Reply by y-kaneda** [reply]
- データベース => 管理下にあるプロセス外依存
- テスト・ダブルに置き換えず直接利用することで退行（regression）に対する保護を得られる

**Comment by y-kaneda** [comment]
## 10.1 データベースをテストするのに必要な事前準備

**Reply by y-kaneda** [reply]
単体テストの時点から必要じゃないですかね・・

**Comment by y-kaneda** [comment]
### 10.1.1 ソースコード管理システムを用いたスキーマの管理

**Reply by y-kaneda** [reply]
- Rails とか Laravel とか差分のスキーマ管理の仕組みあって良い

**Comment by y-kaneda** [comment]
### 10.1.2 スキーマに参照データ（reference data）を含めること

**Reply by y-kaneda** [reply]
- 例えばマスタみたいなデータは(ケースごとに)都度作らないで事前にGitとかにInsert文を保存しておくのが良さそう
  - Rails で言えば seed みたいな

**Comment by y-kaneda** [comment]
### 10.1.3 開発者ごとに個別のデータベース・インスタンスを用意する

**Comment by y-kaneda** [comment]
### 10.1.4 データベースに対する変更の本番環境への反映：状態ベース vs 移行ベース

**Reply by y-kaneda** [reply]
- 状態ベース => 実際に開発用DBに対して行ったスキーマ変更に対して差分を取ってSQLスクリプトを自動生成し、それを本番DBに反映
- 移行ベース => 変更するためのSQLスクリプトを手動作成して反映
  - SQLスクリプトはGit管理

**Comment by y-kaneda** [comment]
## 10.2 データベース・トランザクションの管理

**Comment by y-kaneda** [comment]
### 10.2.1 プロダクション・コードにおけるデータベース・トランザクションの管理

**Reply by y-kaneda** [reply]
- commitメソッドはコントローラーで呼び出す件 -> usecase に読み替えてる
- 単位作業(Unit of Work)
  - ビジネスロジックというか諸々の処理をした後に保持しておいた変更をまとめてDBに反映させる

**Reply by nmurakami0** [reply]
Unit of Work

- 対象のビジネスオペレーションが終わるときに変更のあった全てのデータをデータベースに対して更新する
- DbContext
    - > DbContext の有効期間は、インスタンスが作成された時点で開始し、インスタンスが破棄された時点で終了します。
    - > SaveChanges への 1 回の呼び出しに含まれるすべての変更がトランザクションに適用されます

気になる点

- Unit of Work中に書き込んだ結果のレコードのIDが必要になったらどうするのか
  - そのような状況が起きるのは、そもそもUnit of Workではないということ？

**Comment by y-kaneda** [comment]
### 10.2.2 統合テストにおけるデータベース・トランザクションの管理

**Reply by y-kaneda** [reply]
- テストケースの準備・実行・確認フェーズでは同じトランザクションを使い回さない
  - 本来のプロダクションコードでは分かれているはずなので、実際の動きと乖離してしまう

**Comment by y-kaneda** [comment]
## 10.3 テスト・データのライフ・サイクル

**Reply by y-kaneda** [reply]
- あるテストケースのデータが他のケースに影響を及ぼしてはいけない

**Comment by y-kaneda** [comment]
### 10.3.1 統合テストにおけるテスト・ケースの実行：同時に複数実行すべきか、もしくは、1つずつ実行すべきか？

**Reply by y-kaneda** [reply]
- 他ケースのデータの影響を気にして頑張って複数同時実行するより、1つずつ実行するほうが現実的
  - これってマスタ的なデータとか、同じデータに対してGETのパラメータ変えて何回かテストして取れるものが変わるようなケースとかでもデータ綺麗にしたほうがいいのかしら
    - 10.3.2 の P.351 の TIPS にマスタデータ(参照データ)は残すって書いてあった

**Comment by y-kaneda** [comment]
### 10.3.2 統合テストでのデータの後始末

**Reply by y-kaneda** [reply]
- 推奨されない方法の例が極端・・

**Comment by y-kaneda** [comment]
### 10.3.3 メモリ内（in-memory）データベースの使用に関する問題

**Reply by y-kaneda** [reply]
- 普通のDBと差異があるのもそうだし、プロダクションで使うDBと常に同じ状態(データ構造)に保つためのコストが高そう

**Comment by y-kaneda** [comment]
## 10.4 テスト・コードの再利用

**Comment by y-kaneda** [comment]
### 10.4.1 準備（Arrange）フェーズでのコードの再利用

**Reply by y-kaneda** [reply]
- データ準備をファクトリ・メソッドにする
- オブジェクト・マザー
  - テスト・フィクスチャ（test fixture）の生成を助けるクラスやメソッド
- テスト・データ・ビルダー
- ファクトリ・メソッドはまずテストコードと同じ場所に実装し、重複してきたらヘルパーに移す。基底クラスには置かない

**Reply by nmurakami0** [reply]
- ビジネスロジックに関わらない技術的なコードをプライベートメソッドやヘルパークラスに抽出する
- arrangeフェーズをファクトリメソッドなどに抽出
    - ビルダーパターンだとコードチェーンになって、ボイラープレートが増える
        - デフォルト値引数がある場合はファクトリメソッドの方がわかりやすい

**Comment by y-kaneda** [comment]
### 10.4.2 実行（Act）フェーズでのコードの再利用

**Reply by nmurakami0** [reply]
- actフェーズ
    - e.g., テスト対象システムの機能への呼び出しが移譲されるメソッドの導入

**Comment by y-kaneda** [comment]
### 10.4.3 確認（Assert）フェーズでのコードの再利用

**Reply by y-kaneda** [reply]
- assert をまとめたメソッド

**Reply by nmurakami0** [reply]
- assertフェーズでもヘルパーメソッドは有効
    - e.g., 実行結果となるデータを取得するQueryUserメソッド

**Comment by y-kaneda** [comment]
### 10.4.4 テストの際にデータベース・トランザクションの数が多くなることは問題にならないのか？

**Reply by y-kaneda** [reply]
- テストケースの量によってはトランザクションの重さはけっこう問題になるのでは・・

**Comment by y-kaneda** [comment]
## 10.5 データベースを使ったテストに関するよくある疑問

**Comment by y-kaneda** [comment]
### 10.5.1 読み込みに対してテストをすべきか？

**Reply by y-kaneda** [reply]
- 読み込みのテストはそんなに重要じゃないってマジか・・？
  - 読み込みのテストってどこからどこまでを言ってるんだろう
  - DBからデータを読み込んでくるインフラ層ぐらいの1点だけの話？
  - GET のAPIレスポンスがフロントが期待してる形と違ってヤバいとかは読み込みの話ではない？

**Reply by nmurakami0** [reply]
- 読み込みのバグによるリスクは低い
  - -> リスクが相対的に低いというのはそうかもしれない
- 読み込みについてはドメインモデルが不要 -> 抽象化層がほぼ必要ない
    - データの変更をしないためカプセル化が不要なため

この本の単体テスト的に、単体テストでDBと接続しないということなら、著者の主張通りかもしれない

**Comment by y-kaneda** [comment]
### 10.5.2 リポジトリをテストすべきか？

**Reply by y-kaneda** [reply]
- create or update とか find or create とか delete insert とかリポジトリでやってる可能性がある
  - これは "リポジトリに複雑さが含まれている" のでテスト対象になり得ると思う
  - _`その複雑さを自己完結型のアルゴリズムとして抽出し、そのアルゴリズムだけを検証の対象とすることです`_ がイマイチよく分からなかった

**Reply by nmurakami0** [reply]
- リポジトリはコントローラーに属する
    - あまり複雑にならない
    - プロセス外依存を扱う
- 統合テストがあるなら、リポジトリのテストでそれに追加される価値は少ない
- ORマッパーを使っているなら、リポジトリを直接検証するのではなく、統合テストのシナリオの一部に含めて検証すると良い

リポジトリがfindとかならたしかにテストは不要だろうけど、他にデータベースアクセスに基づくロジックが入りそうな気もする。
そもそもこの本でいう統合テスト的なのに、一般的には単体テストだろうものが含まれるので、
その場合リポジトリのテストは書くような気がする。

**Comment by y-kaneda** [comment]
## 10.6 結論

**Comment by y-kaneda** [comment]
## まとめ
